- Description: 3270: Missing pointer checks.
  Symptom:     Kernel crashes if the subchannel of a 3270 device is detached.
  Problem:     If the subchannel of a 3270 is detached while an i/o is
               in progress the common i/o layer passes an error encoded
               in the irb pointer to the 3270 interrupt handler. The
               interrupt handler needs to check for this special case.
  Solution:    Add missing pointer sanity checks.
  Problem-ID:  -
  Archs:       s390-32, s390-64

--- linux-2.5/drivers/s390/char/raw3270.c	24 Oct 2003 12:10:09 -0000	1.15
+++ linux-2.5/drivers/s390/char/raw3270.c	23 Apr 2004 12:36:20 -0000	1.15.6.1
@@ -345,8 +345,10 @@
 	rq = (struct raw3270_request *) intparm;
 	view = rq ? rq->view : rp->view;
 
-	if (irb->scsw.dstat == 
-	    (DEV_STAT_CHN_END | DEV_STAT_DEV_END | DEV_STAT_UNIT_EXCEP)) {
+	if (IS_ERR(irb))
+		rc = RAW3270_IO_RETRY;
+	else if (irb->scsw.dstat ==  (DEV_STAT_CHN_END | DEV_STAT_DEV_END |
+				      DEV_STAT_UNIT_EXCEP)) {
 		/* Handle CE-DE-UE and subsequent UDE */
 		set_bit(RAW3270_FLAGS_BUSY, &rp->flags);
 		rc = RAW3270_IO_BUSY;
@@ -381,6 +383,8 @@
 			return;	/* Sucessfully restarted. */
 		break;
 	case RAW3270_IO_STOP:
+		if (!rq)
+			break;
 		raw3270_halt_io_nolock(rp, rq);
 		rq->rc = -EIO;
 		break;
@@ -881,7 +885,7 @@
 		if (rc) {
 			/* Didn't work. Try to reactivate the old view. */
 			rp->view = oldview;
-			if (oldview->fn->activate(oldview) != 0) {
+			if (!oldview || oldview->fn->activate(oldview) != 0) {
 				/* Didn't work as well. Try any other view. */
 				list_for_each_entry(nv, &rp->view_list, list)
 					if (nv != view && nv != oldview) {

